File: Source/Core/Mappings/MapTrie.cs
using System;
using System.Collections.Generic;

namespace Quip {
  public class MapTrie {
    class Node {
      Dictionary<Key, Node> children = new Dictionary<Key, Node>();
      MapHandler item;

      public void Insert (Keystroke keystroke, int index, MapHandler handler) {
        if (index == keystroke.Length) {
          item = handler;
        } else {
          var letter = keystroke[index];
          var child = default(Node);
          if (!children.TryGetValue(letter, out child)) {
            child = new Node();
            children[letter] = child;
          }

          child.Insert(keystroke, index + 1, handler);
        }
      }

      public MapHandler Find (Keystroke keystroke, int index) {
        if (index == keystroke.Length) {
          return item;
        } else {
          var letter = keystroke[index];
          var child = default(Node);
          if (children.TryGetValue(letter, out child)) {
            return child.Find(keystroke, index + 1);
          }
        }

        return null;
      }
    }

    public void Insert (Keystroke keystroke, MapHandler handler) {
      m_root.Insert(keystroke, 0, handler); 
    }

    public MapHandler Find (Keystroke keystroke) {
      return m_root.Find(keystroke, 0);
    }

    Node m_root = new Node();
  }
}

File: Code/Core/Modes/MapTrie.cpp
#include "MapTrie.hpp"

#include "KeySequence.hpp"

namespace quip {
  MapTrie::MapTrie ()
  : m_root(std::make_unique<MapTrieNode>()) {
  }
  
  void MapTrie::insert (const KeySequence & sequence, MapHandler handler) {
    const Key * key = sequence.begin();
    const Key * end = sequence.end();
    MapTrieNode * node = m_root.get();
    
    while(key != end) {
      MapTrieNode * child = node->findChild(*key);
      if (child == nullptr) {
        // The remainder of the sequence has no structure in the trie, add it.
        while(key != end) {
          node = node->findOrAddChild(*key);
          ++key;
        }
      } else {
        node = child;
        ++key;
      }
    }
    
    node->setHandler(handler);
  }
  
  const MapTrieNode * MapTrie::find (const KeySequence & sequence) const {
    const Key * key = sequence.begin();
    const Key * end = sequence.end();
    MapTrieNode * node = m_root.get();
    
    while (key != end) {
      node = node->findChild(*key++);
      if (node == nullptr) {
        break;
      }
    }
    
    return node;
  }
}